drivers/gnss/gnss_u_blox_f9p.c

```diff
@@ -0,0 +1,500 @@
+	&disable_gll, &disable_gns, &disable_grs, &disable_gsa, &disable_gst,
+	&disable_vlw, &disable_vtg, &disable_zda, &enable_nav, &nav_fix_mode_auto,
+#if CONFIG_GNSS_SATELLITES
+	&enable_sat,
+#endif
+);
+
+MODEM_UBX_MATCH_ARRAY_DEFINE(u_blox_f9p_unsol_messages,
+	MODEM_UBX_MATCH_DEFINE(UBX_CLASS_ID_NAV, UBX_MSG_ID_NAV_PVT,
+			       gnss_ubx_common_pvt_callback),
+#if CONFIG_GNSS_SATELLITES
+	MODEM_UBX_MATCH_DEFINE(UBX_CLASS_ID_NAV, UBX_MSG_ID_NAV_SAT,
+			       gnss_ubx_common_satellite_callback),
+#endif
+);
+
+static int ubx_f9p_msg_get(const struct device *dev, const struct ubx_frame *req,
+			  size_t len, void *rsp, size_t min_rsp_size)
+{
+	struct ubx_f9p_data *data = dev->data;
+	struct ubx_frame *rsp_frame = (struct ubx_frame *)data->script.inst.response.buf;
+	int err;
+
+	err = k_mutex_lock(&data->script.lock, K_SECONDS(3));
+	if (err != 0) {
+		LOG_ERR("Failed to take script lock: %d", err);
+		return err;
+	}
+
+	data->script.inst.timeout = K_SECONDS(3);
+	data->script.inst.retry_count = 2;
+	data->script.inst.match.filter.class = req->class;
+	data->script.inst.match.filter.id = req->id;
+	data->script.inst.request.buf = req;
+	data->script.inst.request.len = len;
+
+	err = modem_ubx_run_script(&data->ubx.inst, &data->script.inst);
+	if (err != 0 || (data->script.inst.response.buf_len < UBX_FRAME_SZ(min_rsp_size))) {
+		err = -EIO;
+	} else {
+		memcpy(rsp, rsp_frame->payload_and_checksum, min_rsp_size);
+	}
+
+	(void)k_mutex_unlock(&data->script.lock);
+
+	return err;
+}
+
+static int ubx_f9p_msg_send(const struct device *dev, const struct ubx_frame *req,
+			  size_t len, bool wait_for_ack)
+{
+	struct ubx_f9p_data *data = dev->data;
+	int err;
+
+	err = k_mutex_lock(&data->script.lock, K_SECONDS(3));
+	if (err != 0) {
+		LOG_ERR("Failed to take script lock: %d", err);
+		return err;
+	}
+
+	data->script.inst.timeout = K_SECONDS(3);
+	data->script.inst.retry_count = wait_for_ack ? 2 : 0;
+	data->script.inst.match.filter.class = wait_for_ack ? UBX_CLASS_ID_ACK : 0;
+	data->script.inst.match.filter.id = UBX_MSG_ID_ACK;
+	data->script.inst.request.buf = req;
+	data->script.inst.request.len = len;
+
+	err = modem_ubx_run_script(&data->ubx.inst, &data->script.inst);
+
+	(void)k_mutex_unlock(&data->script.lock);
+
+	return err;
+}
+
+static int ubx_f9p_msg_payload_send(const struct device *dev, uint8_t class, uint8_t id,
+				   const uint8_t *payload, size_t payload_size, bool wait_for_ack)
+{
+	struct ubx_f9p_data *data = dev->data;
+	struct ubx_frame *frame = (struct ubx_frame *)data->script.request_buf;
+	int err;
+
+	err = k_mutex_lock(&data->script.lock, K_SECONDS(3));
+	if (err != 0) {
+		LOG_ERR("Failed to take script lock: %d", err);
+		return err;
+	}
+
+	err = ubx_frame_encode(class, id, payload, payload_size,
+			       (uint8_t *)frame, sizeof(data->script.request_buf));
+	if (err > 0) {
+		err = ubx_f9p_msg_send(dev, frame, err, wait_for_ack);
+	}
+
+	(void)k_mutex_unlock(&data->script.lock);
```

Comment:
please use k_sem for locking when thread ownership is not tracked (like would be the case for a condvar) its way cheaper and faster :)

Reply:
Done. FWIW - It was a mutex because `ubx_f9p_msg_payload_send()` would then call `ubx_f9p_msg_send()` and would take advantage of double-locking the same lock. I've decided to split these in two semaphores in order to serve the same purpose.